### PR TITLE
Add nest_string()/unnest_string() functions

### PR DESCRIPTION
**Pull request by aaronwolen**
This PR slightly modifies `unnest()` so the transform step from the example isn't necessary and can just be written as:

```
df <- data.frame(
  x = 1:3,
  y = c("a", "d,e,f", "g,h"),
  stringsAsFactors = FALSE
)
unnest(df, y)
```

I'm often doing these kind of unnesting operations and just wanted to save myself some typing by letting `unnest()` handle the string splitting. 

This also adds a `nest()` function so it's possible to round trip:

```
df %>% unnest(y) %>% nest(y)
```

### CODE REVIEW

**Comment by hadley** [comment]
Hmmmm, that's only one type of nesting - the other is when you have something like:

```R
dplyr::data_frame(x = c(1, 2), y = list(1:3, 9:10))
```

So maybe a more specific name like `unnest_string()` ?

**Comment by aaronwolen** [comment]
Ah, good call. `unnest_string()` makes sense and fits with the precedent set by `extract_numeric()`. 

But which type of nesting do you think is more common? If it's string nesting then maybe `unnest_list()` makes more sense?

**Comment by aaronwolen** [comment]
Thinking about it for the past 15 seconds, `unnest_list()` sounds more like a `base::unsplit()` replacement than a data.frame function. `unnest_string()` is probably the way to go.

Are you envisioning `nest_string()`/`unnest_string()` and `nest()`/`unnest()` for list columns? Or is that nest overload?

**Comment by hadley** [comment]
To me, `unnest()` seems like the atomic operation (in the sense you combine it with other atoms to do something useful), so I'd like to keep it the primary verb.

File: R/nest-string.R [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#' @export
+#' @examples
+#'
+#' df <- data.frame(
+#'   x = 1:3,
+#'   y = c("a", "d,e,f", "g,h"),
+#'   stringsAsFactors = FALSE
+#' )
+#' unnest_string(df, y)
+unnest_string <- function(data, col = NULL, sep = "[^[:alnum:]]+", ...) {
```
Comment (hadley) [inline R/nest-string.R]: Doesn't this always need a col?

File: R/nest-string.R [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#' @export
+#' @examples
+#'
+#' df <- data.frame(
+#'   x = c(1, 2, 2, 2, 3, 3),
+#'   y = c("a", "d", "e", "f", "g", "h"),
+#'   stringsAsFactors = FALSE
+#' )
+#' nest_string(df, y)
+nest_string <- function(data, col = NULL, sep = ",") {
```
Comment (hadley) [inline R/nest-string.R]: I'm not sure I want `nest_string()` - I think `unnest_string()` is useful, but the reverse seems a little ill-specified to me (esp how the grouping work)

**Comment by hadley** [comment]
Could you PTAL at the build failure?

**Comment by aaronwolen** [comment]
Looks like the build failure was introduced in #59. I added a fix here (9c6fcc4) but can submit a separate PR If you prefer.

**Comment by aaronwolen** [comment]
Thanks for the comments. I see your point about `nest_string()` and removed it to wrap up this PR.

Any interest in extending `unnest()` (and `unnest_string()`) to accept multiple columns as suggested in #44?

File: DESCRIPTION [hunk 3]
```diff
@@ -17,6 +17,8 @@ Imports:
     dplyr (>= 0.2),
     stringi,
     lazyeval
+Suggets:
```
Comment (hadley) [inline DESCRIPTION]: Was this deliberate?

**Comment by hadley** [comment]
Would you mind including a couple of unit tests too please?

File: DESCRIPTION (hunk 3 above)
Comment (aaronwolen) [inline DESCRIPTION]: I meant to add data.table to the suggested packages to address this [build failure](https://travis-ci.org/hadley/tidyr/builds/48924616), but listing it under a redundant/misspelled key was just a silly mistake (insert embarrassed emoji here).

Comment (hadley) [inline DESCRIPTION]: Ah I just discovered that too and fixed it (in a slightly different way). Could you please merge/rebase?

**Comment by aaronwolen** [comment]
Sure, I'll rebase and add a few tests.

**Comment by hadley** [comment]
I wonder if this should be `unnest_separate()` since it's similar to separate?

**Comment by aaronwolen** [comment]
Yeah, I think that makes sense. Should I update the PR?

**Comment by hadley** [comment]
Yes, that would be great!  I think maybe it should have `col` and `convert` arguments to be as similar as possible to `separate()`.  And maybe it should be called `separate_rows()`? Or something like that. 

It would also be useful give this function, `extract()`, and `separate()` a common `@family`, maybe `@family string splitting` ?

**Comment by aaronwolen** [comment]
From a readability perspective I definitely prefer `separate_rows()` to `unnest_separate()`. 

However, I think an argument can be made that this function is conceptually more similar to `unnest()`, which produces new rows, than it is to `separate()`, which produces a new column. If you buy that argument, `unnest_*something*()` might be more inline with user expectations, and if you don't... well, then  `separate_rows()` it is!

**Comment by hadley** [comment]
It's half-way between both, but I think people are more likely to look for it after discovering that `separate()` doesn't do what they want. I think if you understand `unnest()` you might not need `unnest_string()`.

File: R/separate-rows.R [hunk 4]
```diff
@@ -57,3 +57,14 @@ separate_rows_.tbl_df <- function(data, cols, sep = "[^[:alnum:].]+",
                                   convert = FALSE) {
   as_data_frame(NextMethod())
 }
+
+#' @export
+separate_rows_.grouped_df <- function(data, cols, sep = "[^[:alnum:].]+",
+                                  convert = FALSE) {
+
```
Comment (aaronwolen) [inline R/separate-rows.R]: Should `separate_rows()` prevent modification of grouped variables, @hadley?

Comment (hadley) [inline R/separate-rows.R]: It shouldn't - you can copy the approach from `separate_`:

```R
#' @export
separate_.grouped_df <- function(data, col, into, sep = "[^[:alnum:]]+",
                                 remove = TRUE, convert = FALSE,
                                 extra = "warn", fill = "warn", ...) {
  regroup(NextMethod(), data, if (remove) col)
}
```

**Comment by aaronwolen** [comment]
Cool, this should be ready for review.

**Comment by hadley** [comment]
Thanks!

**Pull request merged**